Uzair's ChatGPT-aided code changes rebased and revised

mambelli opened the pull request:
I started with Uzar's code in PR #432, squashed the commits preserving the messages, rebased to the current master, and reviewed the code fixing bugs and reintegrating missing parts.

Generally, the AI-aided changes improved the docstrings language and regularized the format, sometimes with inconsistencies file by file (e.g. inconsistent newline after the triple quote).

The AI-aided changes fixed a couple of bugs and introduced behavior differences (e.g. removing return statements) and some bugs. Mostly blatant.  E.g.: it removed a couple of function bodies and changed indentation removing methods from a class.

This PR supersedes PR #432.

namrathaurs commented on frontend/glideinFrontendDowntimeLib.py:
```diff
@@ -16,33 +16,97 @@
+            check_time (int, optional): The time to check against, in UNIX timestamp format.
+                Defaults to the current time.
+
+        Returns:
```
Is this redundant? I haven't seen any occurrences of `Returns:` being described as None when the method does not return anything.

mambelli replied:
The method has a return statement. I added an explanation.

namrathaurs commented on frontend/glideinFrontendDowntimeLib.py:
```diff
@@ -229,6 +332,16 @@ def endDowntime(fname, end_time=None):
+        check_time (int, optional): The time to check against, in UNIX timestamp format.
+            Defaults to the current time.
+
+    Returns:
```
Is this redundant again?

mambelli replied:
removed

namrathaurs commented on lib/cleanupSupport.py:
```diff
@@ -129,13 +188,30 @@ def get_files_wstats(self):
     def delete_file(self, fpath):
+        """Deletes a file from the filesystem.
+
+        This is likely reimplemented by the children
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/condorExe.py:
```diff
@@ -1,9 +1,8 @@
-#   This module implements the functions to execute condor commands
-
+"""This module implements the functions to execute condor commands.
+"""
```
Should this be added at the end of the previous line itself? (looked at other docstrings in this file for reference before adding this comment)

mambelli replied:
done

namrathaurs commented on lib/condorExe.py:
```diff
@@ -13,36 +12,52 @@
 
 class CondorExeError(RuntimeError):
-    """Base class for condorExe module errors"""
+    """Base class for condorExe module errors."""
```
Need to mention inheritance of `RuntimeError` class?

mambelli replied:
no, in class signature

namrathaurs commented on lib/condorExe.py:
```diff
@@ -13,36 +12,52 @@
+    """Set path to condor binaries, if needed.
 
-    Works changing the global variables condor_bin_path and condor_sbin_path
+    Works by changing the global variables condor_bin_path and condor_sbin_path.
```
Should the variables be quoted as `condor_bin_path` and `condor_sbin_path` since they are variable names?

mambelli replied:
done

namrathaurs commented on lib/condorExe.py:
```diff
@@ -96,22 +129,18 @@ def exe_cmd_sbin(condor_exe, args, stdin_data=None, env={}):
 def generate_bash_script(cmd, environment):
-    """Print to a string a shell script setting the environment in 'environment' and running 'cmd'
-    If 'cmd' last argument is a file it will be printed as well in the string
+    """Print to a string a shell script setting the environment in 'environment' and running 'cmd'.
```
Change `'` to `

mambelli replied:
done

namrathaurs commented on lib/condorExe.py:
```diff
@@ -96,22 +129,18 @@ def exe_cmd_sbin(condor_exe, args, stdin_data=None, env={}):
-    If 'cmd' last argument is a file it will be printed as well in the string
+    """Print to a string a shell script setting the environment in 'environment' and running 'cmd'.
+
+    If 'cmd' last argument is a file it will be printed as well in the string.
```
Same comment as above

mambelli replied:
done

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
 ##############################################
 # Helper functions of condorAdvertise
 def usetcp2str(use_tcp):
+    """Convert use_tcp flag to a string suitable for the Condor command line.
```
Enclose `use_tcp` in ` to denote that it is a variable

mambelli replied:
done

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        use_tcp (bool): If True, use TCP.
+
+    Returns:
+        str: The use_tcp flag formatted for the command line.
```
Enclose `use_tcp` in ` to denote that it is a variable

mambelli replied:
done

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
 
 
 def ismulti2str(is_multi):
+    """Convert is_multi flag to a string suitable for the Condor command line.
```
Enclose `is_multi` in ` to denote that it is a variable

mambelli replied:
done

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        is_multi (bool): If True, indicate multiple.
+
+    Returns:
+        str: The is_multi flag formatted for the command line.
```
Enclose `is_multi` in ` to denote that it is a variable

mambelli replied:
done

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        pool_name (str, optional): The name of the pool. Defaults to None.
+
+    Returns:
+        str: The output of the condor_advertise command.
```
Should `condor_advertise` be enclosed in ` ?

mambelli replied:
It could be either way, it is not in the code. Anyway done

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
-# Issue a condor_reschedule
-#
 def condorReschedule(schedd_name=None, pool_name=None, schedd_lookup_cache=condorMonitor.local_schedd_cache):
+    """Issue a condor_reschedule command.
```
Should `condor_reschedule` be enclosed in ` ?

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        schedd_lookup_cache (optional): The cache for schedd lookups. Defaults to condorMonitor.local_schedd_cache.
+
+    Returns:
+        str: The output of the condor_release command.
```
Should `condor_release` be enclosed in ` ?

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        schedd_lookup_cache (optional): The cache for schedd lookups. Defaults to condorMonitor.local_schedd_cache.
+
+    Returns:
+        str: The output of the condor_release command.
```
Should `condor_release` be enclosed in ` ?

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        schedd_lookup_cache (optional): The cache for schedd lookups. Defaults to condorMonitor.local_schedd_cache.
+
+    Returns:
+        str: The output of the condor_hold command.
```
Should `condor_hold` be enclosed in ` ?

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        schedd_lookup_cache (optional): The cache for schedd lookups. Defaults to condorMonitor.local_schedd_cache.
+
+    Returns:
+        str: The output of the condor_hold command.
```
Should `condor_hold` be enclosed in ` ?

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        schedd_lookup_cache (optional): The cache for schedd lookups. Defaults to condorMonitor.local_schedd_cache.
+
+    Returns:
+        str: The output of the condor_rm command.
```
Should `condor_rm` be enclosed in ` ?

namrathaurs commented on lib/condorManager.py:
```diff
@@ -56,105 +91,182 @@ def condorSubmitOne(
+        schedd_lookup_cache (optional): The cache for schedd lookups. Defaults to condorMonitor.local_schedd_cache.
+
+    Returns:
+        str: The output of the condor_rm command.
```
Should `condor_rm` be enclosed in ` ?

namrathaurs commented on lib/condorSecurity.py:
```diff
@@ -60,17 +68,36 @@ def load(self):
+
+        Args:
+            sec_filter (dict): Dictionary of security contexts and features.
+                               Example: `[contex]=[feature list]`
```
Should it be `context`? Is there a typo?

mambelli replied:
fixed, context

namrathaurs commented on lib/condorSecurity.py:
```diff
@@ -60,17 +68,36 @@ def load(self):
+    return filter_list
 
 
 class SecEnvState(EnvState):
```
Should the inherited class be mentioned in the docstring?

mambelli replied:
it's in the signature

namrathaurs commented on lib/condorSecurity.py:
```diff
@@ -189,12 +255,23 @@ def enforce_requests(self):
 
 ########################################
 class EnvProtoState(SecEnvState):
```
Should the inherited class be mentioned in the docstring?

mambelli replied:
If you mean the parent, it's in the signature

namrathaurs commented on lib/condorSecurity.py:
```diff
@@ -215,7 +292,25 @@ def __init__(self, filter=None):
 # the context and feature are related
 # to the Condor protocol handling
 class ProtoRequest(SecEnvRequest):
```
Should the inherited class be mentioned in the docstring?

mambelli replied:
If you mean the parent, it's in the signature

namrathaurs commented on lib/condorSecurity.py:
```diff
@@ -242,7 +349,27 @@ def get(self, context, feature):
 
 
 class GSIRequest(ProtoRequest):
```
Should the inherited class be mentioned in the docstring?

mambelli replied:
If you mean the parent, it's in the signature

namrathaurs reviewed (commented):
Part 1 of PR review completed with 17/30 files reviewed with minor comments.

mambelli commented on factory/glideFactoryLogParser.py:
```diff
@@ -187,7 +187,7 @@ def diff_raw(self, other):
-    def diff(self, other):
+    def diff(self, other_data):
         """
         Diff self.data with other for use in comparing current
```
done

mambelli commented on lib/exprParser.py:
```diff
@@ -2,7 +2,7 @@
 
 """
-Description: general purpose python expression parser and unparser
+Description: General purpose Python expression parser and unparser.
```
done

mambelli commented on lib/glideinWMSVersion.py:
```diff
@@ -20,10 +20,16 @@
+        """Implementation of the singleton interface."""
 
         def __init__(self, chksumFile="checksum"):
+            """
```
It is not adding much but I did it

mambelli commented on lib/glideinWMSVersion.py:
```diff
@@ -74,57 +81,66 @@ def createVersionString(self):
 
     def __getattr__(self, attr):
-        """Delegate access to implementation"""
+        """Delegate access to implementation."""
```
I would not do it for getters and setters. Docstrings are together w/ the code.
"Attribute to get" would not add much

mambelli commented on lib/glideinWMSVersion.py:
```diff
@@ -74,57 +81,66 @@ def createVersionString(self):
 
     def __setattr__(self, attr, value):
-        """Delegate access to implementation"""
+        """Delegate access to implementation."""
```
same as 114

mambelli commented on lib/pidSupport.py:
```diff
@@ -13,36 +13,68 @@
-# this exception is raised when trying to register a pid
-# but another process is already owning the PID file
 class AlreadyRunning(RuntimeError):
+    """Exception raised during PID registration when a process is already running and owns the PID file."""
```
Docstrings are with the code (at least with the headers). "Inherits from RuntimeError" is restating the same as "class AlreadyRunning(RuntimeError)"

mambelli commented on lib/servicePerformance.py:
```diff
@@ -3,63 +3,96 @@
+        """Initialize a PerfMetric object.
+
+        Args:
+            name (str): The name of the service.
```
No, `def __init__(self, name)` has no `metric`.

namrathaurs replied:
Looks like I got confused seeing `self.metric` being used in the `__init__` method. Can be resolved.

mambelli commented on lib/subprocessSupport.py:
```diff
@@ -18,42 +17,40 @@
     Returns:
-        str/bytes: output of the command. It will be bytes if text is False,
-            str otherwise
+        str or bytes: The output of the command. The type depends on the value of `text`.
```
I thought the explanation "The type depends on the value of `text`." and the text definition "text (bool): Whether to treat stdin, stdout, and stderr as text (str) or bytes. Defaults to True." would be clear enough.
What would you add?

namrathaurs replied:
I see, what you said makes sense to me now. I was thinking more along the lines of clarifying what True actually represents. From what I read, when treated as `str`, `text` would be set to `True`?

namrathaurs replied:
Addressed

mambelli commented on lib/subprocessSupport.py:
```diff
@@ -18,42 +17,40 @@
     Raises:
-        subprocess.CalledProcessError: if the subprocess fails (exit status not 0)
-        RuntimeError: if it fails to invoke the subprocess or the subprocess times out
+        subprocess.CalledProcessError: If the command returns a non-zero exit code.
```
`subprocess.CalledProcessError` is the type, there is no module `command`

mambelli commented on lib/util.py:
```diff
@@ -45,10 +45,13 @@
 def flattenDict(d, join=add, lift=lambda x: x):
     """Flexible flattening of a dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -99,15 +102,19 @@ def dict_to_flat(in_dict, prefix="", suffix="", sep=""):
+    Notes:
+        Value could be clobbered if there are duplicates in the strings resulting from concatenating keys, e.g.
+        `{'a':{'b':1}, 'ab':2}`
+        A separator will not solve the problem if it is a valid character for the keys
```
done

mambelli commented on lib/util.py:
```diff
@@ -45,10 +45,13 @@
-    :param lift: lift function for the keys (allows to create tuples)
-    :return: list with flattened dictionary
+    Args:
+        d (dict): Dictionary to flatten
```
done

mambelli commented on lib/util.py:
```diff
@@ -45,10 +45,13 @@
+        lift (function): lift function for the keys (allows to create tuples). Default is identity.
+
+    Returns:
+        list: List with flattened dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -99,15 +102,19 @@ def dict_to_flat(in_dict, prefix="", suffix="", sep=""):
+        A separator will not solve the problem if it is a valid character for the keys
+
+    Args:
+        in_dict (dict): Input dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -99,15 +102,19 @@ def dict_to_flat(in_dict, prefix="", suffix="", sep=""):
-    :param sep: separator between keys (Default: "")
-    :return: flattened dictionary
+    Returns:
+        str: Flattened dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -125,10 +132,13 @@ def dict_to_flat_slow(in_dict, prefix="", suffix=""):
     """Flatten a multi-level dictionary to one level
     The resulting keys are the string concatenation of the original ones
```
done

mambelli commented on lib/util.py:
```diff
@@ -125,10 +132,13 @@ def dict_to_flat_slow(in_dict, prefix="", suffix=""):
     """Flatten a multi-level dictionary to one level
```
done

mambelli commented on lib/util.py:
```diff
@@ -125,10 +132,13 @@ def dict_to_flat_slow(in_dict, prefix="", suffix=""):
-    :param suffix: suffix for the keys (Default "")
-    :return: flattened dictionary
+    Args:
+        in_dict (dict): Input dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -125,10 +132,13 @@ def dict_to_flat_slow(in_dict, prefix="", suffix=""):
+        suffix (str, optional): Suffix for the keys. Default is "".
+
+    Returns:
+        str: Flattened dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -143,12 +153,16 @@ def dict_to_flat_slow(in_dict, prefix="", suffix=""):
 def dict_normalize(in_dict, keys=None, prefix="", suffix="", default=None):
     """Change the keys of a dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -143,12 +153,16 @@ def dict_to_flat_slow(in_dict, prefix="", suffix=""):
-    :param default: default value passed to get (Default: None)
-    :return: normalized dictionary
+    Args:
+        in_dict (dict): Input dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -143,12 +153,16 @@ def dict_to_flat_slow(in_dict, prefix="", suffix=""):
+        default (Object, optional): Default value passed to get(). Default is None.
+
+    Returns:
+        dict: Normalized dictionary
```
done

mambelli commented on lib/util.py:
```diff
@@ -188,9 +202,9 @@ class ExpiredFileException(Exception):
 def print_funct(*args, **kwargs):
     """Print function that can be used as mask exception (the print statement cannot be passed as parameter)
```
done

mambelli commented on lib/util.py:
```diff
@@ -188,9 +202,9 @@ class ExpiredFileException(Exception):
-    :param kwargs: keywords, valid keywords: 'sep' separator, default is space
-    :return: None
+    Args:
+        *args: list of what to print, converted into string and separated by `sep`
```
done

mambelli commented on lib/util.py:
```diff
@@ -369,17 +383,22 @@ def file_tmp2final(
 # may have to be moved to a shared place (/tmp) and handled more properly
 def file_get_tmp(fname=None, tmp_type=None):
     """Get the name of a temporary file
```
done

mambelli replied:
done

mambelli commented on lib/util.py:
```diff
@@ -369,17 +383,22 @@ def file_tmp2final(
+    `tempfile` from `tempfile.mkstemp()` that guarantees uniqueness and safety
+
+    Args:
+        fname (str, optional): Original file name
```
done

mambelli commented on frontend/glideinFrontendDowntimeLib.py:
```diff
@@ -16,33 +16,97 @@
+    """
+
     def __init__(self, fname):
+        """
```
done

mambelli commented:
I addressed all the comments, I followed the ones resolved, and left open the ones I disagree with. You can resolve them or we can discuss.

namrathaurs commented on lib/condorMonitor.py:
```diff
@@ -779,53 +881,100 @@ def __init__(self, query, group_key_func, group_element_func=None):
+    """Class summarizing other classes.
+
+    Attributes:
+        query (object): The query to summarize
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/condorMonitor.py:
```diff
@@ -839,9 +988,17 @@ def getHash(self, hash_func):
-# check that req_format_els are present in in_format_list, and if not add them
-# return a new format_list
 def complete_format_list(in_format_list, req_format_els):
+    """
```
Append the text of the docstring after `"""`

mambelli replied:
done

namrathaurs commented on lib/condorMonitor.py:
```diff
@@ -1207,12 +1442,15 @@ def fetch2count(data, hash_func):
 def fetch2count_flat(data, hash_func):
     """Count the hash values returned from all the elements in data
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/condorMonitor.py:
```diff
@@ -1207,12 +1442,15 @@ def fetch2count(data, hash_func):
+        data: data from a fetch()
+        hash_func: Hashing function
+                   One argument: classad dictionary
+                   Returns: flat hash value (for hashing functions returning also lists, use fetch2count
+                            if None, will not be counted
```
The parantheses has not been closed

mambelli replied:
fixed

namrathaurs commented on lib/condorMonitor.py:
```diff
@@ -1207,12 +1442,15 @@ def fetch2count(data, hash_func):
+                            if None, will not be counted
+
+    Returns:
+        dict: a dictionary with a count of the hash values returned
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs reviewed (commented):
Part 2 of PR review completed with 25/30 files reviewed with comments.

mambelli commented on lib/hashCrypto.py:
```diff
@@ -27,171 +20,252 @@
+# Explicit hash algorithms section
 
 
 class HashMD5(Hash):
```
No, is in signature

mambelli commented on lib/token_util.py:
```diff
@@ -73,27 +78,27 @@ def token_str_expired(token_str):
+    """Performs a simple scramble (XOR) on a binary string using HTCondor's algorithm.
 
-    simply XOR with 0xdeadbeef
+    Simply XOR with 0xdeadbeef
```
yes, it is just a scrambling to add noise and is also in the HTCondor source in github

mambelli commented on lib/tarSupport.py:
```diff
@@ -7,80 +7,83 @@
+        full_path (str): The full path to the missing file.
+
+    Notes:
+        Must include the file name in `full_path`
```
done

mambelli commented on frontend/glideinFrontend.py:
```diff
@@ -254,7 +322,21 @@ def spawn_iteration(work_dir, frontendDescript, groups, max_active, failure_dict
+        frontend_name (str): The name of the frontend.
+        ha_mode (str): High-availability mode.
+
+    Returns:
```
removed

mambelli commented on frontend/glideinFrontend.py:
```diff
@@ -301,6 +383,24 @@ def spawn(
+        restart_interval (int): Interval (in seconds) before attempting a restart.
+        restart_attempts (int): Maximum number of restart attempts.
+
+    Returns:
```
removed

mambelli commented on frontend/glideinFrontend.py:
```diff
@@ -442,19 +550,43 @@ def shouldHibernate(frontendDescript, work_dir, ha, mode, groups):
+        frontendDescript (FrontendDescript): The frontend descriptor object.
+        element (Element, optional): The specific group element. Defaults to None.
+
+    Returns:
```
done

mambelli commented on frontend/glideinFrontend.py:
```diff
@@ -468,11 +600,27 @@ def set_frontend_htcondor_env(work_dir, frontendDescript, element=None):
+    Args:
+        env (dict): Dictionary of environment variables and their values.
+
+    Returns:
```
done

mambelli commented on frontend/glideinFrontend.py:
```diff
@@ -482,50 +630,80 @@ def clean_htcondor_env():
+        max_parallel_workers (int): Maximum number of parallel workers.
+        removal_action (str): The specific removal action to perform.
+
+    Returns:
```
done

mambelli commented on frontend/glideinFrontend.py:
```diff
@@ -482,50 +630,80 @@ def clean_htcondor_env():
+    Removes environment variables related to CONDOR and X509 to ensure
+    a clean execution environment.
+
+    Returns:
```
done

mambelli commented on lib/condorLogParser.py:
```diff
@@ -89,91 +98,115 @@ def load(self):
-        Merge self data with other info
-        @return: merged data, may modify other
+    def merge(self, other_data):
+        """Merge self data with other data.
```
I think it is correct. The merge is with other data that is in `other_data`, not with `other_data` data. `self.data` and `other_data`. The name was changed from `other` to `other_data` to make that more clear.

mambelli commented on lib/condorLogParser.py:
```diff
@@ -89,91 +98,115 @@ def load(self):
-        diff self data with other info.  Used to compare
-        previous iteration with current iteration
+    def diff(self, other_data):
+        """Compare self data with other data.
```
same

mambelli commented on lib/condorLogParser.py:
```diff
@@ -234,52 +268,65 @@ def loadFromLog(self):
-        Merge self data with other info
-        @return: merged data, may modify other
+    def merge(self, other_data):
+        """Merge self data with other data.
```
same

mambelli commented on lib/condorLogParser.py:
```diff
@@ -234,52 +268,65 @@ def loadFromLog(self):
-        Diff self.data with other info.
-        For use in comparing previous iteration with current iteration
+    def diff(self, other_data):
+        """Compare self data with other data.
```
same

mambelli commented on lib/condorLogParser.py:
```diff
@@ -316,67 +363,88 @@ def diff(self, other):
-        Merge self data with other info
-        @return: merged data, may modify other
+    def merge(self, other_data):
+        """Merge self data with other data.
```
same

mambelli commented on lib/condorLogParser.py:
```diff
@@ -316,67 +363,88 @@ def diff(self, other):
-        Diff self data with other info
-        @return: diff of counts
+    def diff(self, other_data):
+        """Compare self data with other data.
```
same

mambelli commented on lib/condorLogParser.py:
```diff
@@ -396,70 +464,91 @@ def diff(self, other):
-        merge self data with other info
-        @return: merged data, may modify other
+    def merge(self, other_data):
+        """Merge self data with other data.
```
same

mambelli commented on lib/condorLogParser.py:
```diff
@@ -662,41 +765,40 @@ def load(self, active_only=True):
+            other_data (dict): The data to compare with.
+
+        Returns:
+            dict: The differences between self.data and other.
```
done and quoted

mambelli commented on lib/condorMonitor.py:
```diff
@@ -39,8 +39,17 @@
+    `CONDOR_CONFIG` environment variable and manually adds `_CONDOR_` prefixed environment variables to
+    the HTCondor parameters.
+
+    Returns:
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -56,8 +65,15 @@ def htcondor_full_reload():
+    Args:
+        new_condor_bin_path (str): The new path to the Condor binaries.
+
+    Returns:
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -68,18 +84,14 @@ def set_path(new_condor_bin_path):
-    """
-    Thrown when there are exceptions using htcondor python bindings or commands
-    """
+    """Thrown when there are exceptions using htcondor python bindings or commands"""
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -68,18 +84,14 @@ def set_path(new_condor_bin_path):
-    """
-    Thrown when there are exceptions using htcondor python bindings
-    """
+    """Thrown when there are exceptions using htcondor python bindings"""
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -235,9 +282,7 @@ def save(self, objid, obj):
-    """
-    Retrieves a list of a single item from the all the factory queues.
-    """
+    """Retrieves a list of a single item from the all the factory queues."""
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
-        """
-        Fetch the classad attributes specified in the format_list
-        matching the constraint
+        """Fetch the classad attributes specified in the format_list matching the constraint
```
done. Changed to "format list".

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
-        matching the constraint
+        """Fetch the classad attributes specified in the format_list matching the constraint
+
+        Args:
```
completed

mambelli commented on lib/condorMonitor.py:
```diff
@@ -120,44 +140,56 @@ def iGetCmdScheddStr(self, schedd_name):
-    Schedd caching class
-    The schedd can be found either through -name attr or through the
-    local disk lookup. Remember which one to use.
+    """Schedd caching class
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
+        Returns:
+            dict: Dict containing classad attributes and values
         """
         raise NotImplementedError("Fetch not implemented")
```
No, it is not in the normal function behavior, only because of the abstract class

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
-        """
-        Fetch the data and store it in self.stored_data
-        """
+        """Fetch the data and store it in self.stored_data"""
```
added

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
-    """
-    Virtual class that implements fetchStored
-    """
+    """Abstract (virtual) class that implements fetchStored"""
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
     stored_data = {}
 
     def fetchStored(self, constraint_func=None):
         """
```
added

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
-        @rtype: dict
-        @return: Same as fetch(), but limited to constraint_func(el)==True
+        Args:
+            constraint_func (function, optional): A boolean function, with only one argument (data el).
```
Added defaults to. Data el is to describe the function.

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
-    Fully implemented class that executes condorq_edit commands. Only provides a method to do bulk
-    updates of jobs using transaction-based API and the condor python bindings. Cannot be used without
-    python bindings
+    """Fully implemented class that executes condorq_edit commands. Only provides a method to do bulk
```
Usually used for Python variables, ...

mambelli commented on lib/condorMonitor.py:
```diff
@@ -268,62 +313,59 @@ def condorq_attrs(q_constraint, attribute_list):
+    Cannot be used without HTCondor Python bindings
     """
 
     def __init__(self, pool_name=None, schedd_name=None):
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -546,10 +599,12 @@ def fetch_using_bindings(self, constraint=None, format_list=None):
         """Fetch the results using htcondor-python bindings
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -497,10 +548,12 @@ def fetch_using_exe(self, constraint=None, format_list=None):
         """Return the results obtained from executing the HTCondor query command
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -546,10 +599,12 @@ def fetch_using_bindings(self, constraint=None, format_list=None):
+            dict: Dictionary containing the results
 
         Raises:
             NotImplementedError: the operation is not implemented using bindings
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -645,9 +725,7 @@ def fetch_using_bindings(self, constraint=None, format_list=None):
-    """
-    Class to implement condor_status. Uses htcondor-python bindings if possible.
-    """
+    """Class to implement condor_status. Uses htcondor-python bindings if possible."""
```
Usually used for Python variables, ...

mambelli commented on lib/condorMonitor.py:
```diff
@@ -700,15 +776,27 @@ def fetch_using_bindings(self, constraint=None, format_list=None):
-    """
-    Virtual class that implements fetch for StoredQuery()
-    """
+    """Abstract (virtual) class that implements fetch for StoredQuery()"""
```
added

mambelli commented on lib/condorMonitor.py:
```diff
@@ -741,8 +845,7 @@ def __repr__(self):
 class Group(BaseSubQuery):
-    """
-    Sub Query class with grouping functionality
+    """Sub Query class with grouping functionality
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -759,8 +862,7 @@ def __init__(self, query, group_key_func, group_data_func):
 class NestedGroup(BaseSubQuery):
-    """
-    Sub Query class with grouping functionality to create nested results
+    """Sub Query class with grouping functionality to create nested results
```
done

mambelli commented on lib/condorMonitor.py:
```diff
@@ -741,8 +845,7 @@ def __repr__(self):
-    """
-    Sub Query class with grouping functionality
+    """Sub Query class with grouping functionality
     Each element has a value that is the summary of the values in a group
```
done

mambelli commented:
I addressed all the comments in part 2, I followed the ones resolved, and left open the ones I disagree with. You can resolve them or we can discuss.

namrathaurs commented on lib/pubCrypto.py:
```diff
@@ -182,34 +191,32 @@ def load(self, key_str=None, key_fname=None):
+        """Load the key into the object from a BIO.
 
-        Protected, overridden by child classes. Used by load
+        Protected method, overridden by child classes. Used by load
```
End the sentence with a period `.`

namrathaurs commented on lib/pubCrypto.py:
```diff
@@ -221,96 +228,126 @@ def save(self, key_fname):
-        """Save the key from the object
+        """Save the RSA key to a BIO object.
 
         Protected, overridden by child classes. Used by save and get
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/pubCrypto.py:
```diff
@@ -221,96 +228,126 @@ def save(self, key_fname):
+            int: Status code returned by `M2Crypto.m2.rsa_write_pub_key`.
+
         Raises:
             KeyError: if the key is not defined
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/pubCrypto.py:
```diff
@@ -221,96 +228,126 @@ def save(self, key_fname):
-                len(data) must be less than len(key)
+            data (str | bytes): The data to encrypt.
+                                bytes-like or str. If unicode, it is encoded using utf-8 before being encrypted.
+                                len(data) must be less than len(key)
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/pubCrypto.py:
```diff
@@ -221,96 +228,126 @@ def save(self, key_fname):
+        Args:
+            data (str | bytes): The data to encrypt.
+                                bytes-like or str. If unicode, it is encoded using utf-8 before being encrypted.
+                                len(data) must be less than len(key)
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/pubCrypto.py:
```diff
@@ -221,96 +228,126 @@ def save(self, key_fname):
+        Args:
+            data (str | bytes): The data to encrypt.
+                                bytes-like or str. If unicode, it is encoded using utf-8 before being encrypted.
+                                len(data) must be less than len(key)
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/pubCrypto.py:
```diff
@@ -343,104 +389,128 @@ def PubRSAKey(self):
-    # generate key function
     def new(self, key_length=None, exponent=65537):
-        """Refresh/Generate a new key and store it in the object
+        """Generate or refresh a new RSA key and store it in the object
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/pubCrypto.py:
```diff
@@ -343,104 +389,128 @@ def PubRSAKey(self):
-                len(data) must be less than len(key)
+            data (str | bytes): The data to sign.
+                                If unicode, it is encoded using utf-8 before being encrypted.
+                                len(data) must be less than len(key)
```
End the sentence with a period `.`

mambelli replied:
done

namrathaurs commented on lib/fork.py:
```diff
@@ -84,24 +90,30 @@ def add(i, j): return i+j
+
+    Can raise OSError and FetchError.
+    Other errors can come from `os.read` and `pickle.load` (but are caught here):
+      - EOFError if the forked process failed an nothing was written to the pipe, if cPickle finds an empty string
```
Typo in `an` (should it be `and`)?

mambelli replied:
done

namrathaurs commented on lib/fork.py:
```diff
@@ -84,24 +90,30 @@ def add(i, j): return i+j
+
+    Raises:
+        FetchError: If an os.read error was encountered or if the forked process failed.
+        OSError: If Bad file descriptor or file already closed or if waitpid syscall returns -1
+        EOFError: If the forked process failed an nothing was written to the pipe, if cPickle finds an empty string
+        IOError: Failure for I/O-related reason, E.g., "pipe file not found" or "disk full"
+        OSError: Other system-related error (includes both former OSError and IOError since Py3.4)
+        pickle.UnpicklingError: Incomplete pickled data
```
Should all these items end with a period `.` (the first one does)?

mambelli replied:
done

namrathaurs reviewed (commented):
Part 3 of PR review completed with 30/30 files reviewed with comments.

mambelli commented on lib/symCrypto.py:
```diff
@@ -197,28 +180,27 @@ def encrypt(self, data):
 
     def encrypt_hex(self, data):
-        """like encrypt, but the result is hex encoded"""
+        """Encrypts data and returns the result as a hex-encoded string."""
```
added

mambelli commented on lib/symCrypto.py:
```diff
@@ -255,91 +236,105 @@ def decrypt_hex(self, data):
         Raises:
-            ValueError: if the format of the code is incorrect
-
+            ValueError: If the format of key_iv_code is incorrect.
```
done

mambelli commented on lib/rrdSupport.py:
```diff
@@ -586,35 +625,53 @@ def info(self, *args):
         Output is a list of lines, as returned from rrdtool dump.
-        This is different from the `dump` method provided by the `rrdtool` package (Python binding)
+        This is different from the `dump` method provided by the `rrdtool` package (Python bindings)
         which outputs to a file or stdout
```
done

mambelli commented on lib/rrdSupport.py:
```diff
@@ -586,35 +625,53 @@ def info(self, *args):
 
         Args:
-            *args: rrdtool dump arguments, joined in single string for the command line
+            *args: Arguments for the rrdtool dump command, joined in single string for the command line
```
done

mambelli commented on lib/rrdSupport.py:
```diff
@@ -627,114 +684,124 @@ def fetch(self, *args):
+    if `fix_rrd is True`, will attempt to dump out rrd to xml, add the missing attributes, then restore.
+    Original file is backed up with time stamp if `backup is True`, obliterated otherwise.
+
+    This function is used by verifyRRD (in Factory and Frontend), invoked during reconfig/upgrade
```
done

mambelli commented on lib/rrdSupport.py:
```diff
@@ -627,114 +684,124 @@ def fetch(self, *args):
+    Original file is backed up with time stamp if `backup is True`, obliterated otherwise.
+
+    This function is used by verifyRRD (in Factory and Frontend), invoked during reconfig/upgrade
+    No logging available, output is to stdout/err
```
done

mambelli commented on lib/rrdSupport.py:
I checked/changed the changes. Will be ok.

mambelli commented on lib/logSupport.py:
```diff
@@ -60,44 +60,51 @@
+    in seconds but the values in the constructor (`maxDays`, `minDays`) are in days
     (24 hour periods) and can be fractions (float).
-
     And the size is measured in Bytes (MBytes in the constructor parameter can be
```
done

mambelli commented on lib/logSupport.py:
```diff
@@ -60,44 +60,51 @@
 
     Attributes:
         filename (str|Path): The full path of the log file. Includes file name. Passed to BaseRotatingHandler
-        interval (int): Max number of seconds before file rotation. 0 to disable timed rotation
+        interval (int): Time interval in seconds before rotation. 0 to disable timed rotation
         min_lifetime (int): Minimum number of seconds before file rotation (used with maxBytes)
-        maxBytes (int): Maximum size of the logfile in bytes before file rotation (used with min days)
+        maxBytes (int): Maximum file size in bytes before rotation. (used with min days)
             0 to disable file size rotation
```
done

mambelli commented on lib/logSupport.py:
```diff
@@ -60,44 +60,51 @@
+        compression (str): Compression format (gz, zip, depending on available compression modules) used for log files.
+        suffix (str): Suffix format for the rotated files.
+        extMatch (re.Pattern): Regex pattern to match the suffix of the rotated files.
+        rolloverAt (int): Time of the next time-based rollover in seconds from Epoch. 0 to disable
```
done

mambelli commented on lib/logSupport.py:
```diff
@@ -110,10 +117,10 @@ def __init__(self, filename, maxDays=1.0, minDays=0.0, maxMBytes=10.0, backupCou
             maxDays (float): Max number of days before file rotation (fraction of day accepted, used in unit test).
                 0 to disable timed rotation
             minDays (float): Minimum number of days before file rotation (used with max MBytes)
-            maxMBytes (float): Maximum size of the logfile in MB before file rotation (used with min days)
+            maxMBytes (float): Maximum file size in MB before rotation. (used with min days)
                 0 to disable file size rotation
```
done

mambelli commented on lib/logSupport.py:
```diff
@@ -294,49 +302,48 @@ def doRollover(self):
-                with open(dfn, "rb") as f_in:
-                    f_out.writelines(f_in)
-                f_out.close()
+                with gzip.open(dfn + ".gz", "wb") as f_out:
+                    with open(dfn, "rb") as f_in:
+                        f_out.writelines(f_in)
```
yes

mambelli commented on lib/logSupport.py:
```diff
@@ -294,49 +302,48 @@ def doRollover(self):
-        log_file_name (str): log file name (same as the logger name)
-        log_dir (str|Path): log directory
-        msg_types (str): log levels to include (comma separated list). Keywords are:
+        log_file_name (str): Log file name (same as the logger name)
+        log_dir (str|Path): Log directory
```
done

mambelli commented on lib/logSupport.py:
```diff
@@ -294,49 +302,48 @@ def doRollover(self):
             Default INFO,WARN,ERR set at the configuration level
-        extension (str): file name extension
-        maxDays (float): Max number of days before file rotation (fraction of day accepted, used in unit test)
+        extension (str): File name extension
+        maxDays (float): Maximum number of days before file rotation (fraction of day accepted, used in unit test)
         minDays (float): Minimum number of days before file rotation (used with max MBytes)
         maxMBytes (float): Maximum size of the logfile in MB before file rotation (used with min days)
         backupCount (int): Number of backups to keep
         compression (str): Compression to use (gz, zip, depending on available compression modules)
```
done

mambelli commented on lib/logSupport.py:
```diff
@@ -384,30 +390,47 @@ def get_processlog_handler(
-        unformated_dict (dict): The dictionary to be formatted for logging
-        log_format (str): format string for logging
+        unformated_dict (dict): The dictionary to be formatted for logging.
+        log_format (str): Format string for logging.
```
added

mambelli commented on lib/logSupport.py:
```diff
@@ -499,19 +537,19 @@ def get_structlog_logger(name):
 
 def get_logger_with_handlers(name, directory, config_data, level=logging.DEBUG):
-    """Create/retrieve a logger, set the handlers, set the starting logging level, and return the logger
+    """Creates or retrieves a logger, sets the handlers, sets the starting logging level, and returns the logger
```
done

mambelli commented on lib/pubCrypto.py:
```diff
@@ -29,28 +28,36 @@
+        prompt1 (str): Prompt for entering the passphrase.
+        prompt2 (str): Prompt for verifying the passphrase.
+
+    Returns:
+        None
```
Fixed "        None (Optional return, Passphrase or None)"

mambelli commented on lib/pubCrypto.py:
```diff
@@ -127,11 +136,14 @@ def __init__(
-            encryption_padding:
-            sign_algo (str): valid signing algorithm (default: 'sha256')
+            key_str (str | bytes, optional): Base64 encoded key as a string or bytes.
+                                             Must be bytes-like object or ASCII string, like base64 inputs
```
done

mambelli commented:
Addressed also part3

namrathaurs reviewed (approved):
Requested changes either resolved/clarified. Changes ready to be merged!
